### PR TITLE
#MayMon(make_chocolate)

### PR DESCRIPTION
**Pull request by MayMon-T3807**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [ ] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by PyaeLinn01** [comment]
Well done.

**Review by PyaeLinn01 — APPROVED** [review]

**Comment by MayMon-T3807** [comment]
> Well done. 

thank you.

File: solutions/make_chocolate.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+"""
```
Comment (PyaeLinn01) [inline solutions/make_chocolate.py:1]: The module docstring is clear, but consider specifying the function's edge cases in more detail, such as when goal = 0

Comment (MayMon-T3807) [reply solutions/make_chocolate.py:1]: yes. thanks.

File: solutions/make_chocolate.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+This module provides a function for determining the number of small chocolate bars needed
+to meet a specific goal weight, prioritizing the use of big bars.
+
+The function calculates the required small bars after using as many big bars as possible,
+ensuring the goal is achievable. It raises appropriate errors for invalid inputs.
+
+@author: May Mon Thant
+"""
+
+
```
Comment (PyaeLinn01) [inline solutions/make_chocolate.py:11]: The function's behavior is well-described. However, the examples could include additional edge cases, such as when small = 0, big = 0, or goal = 0

File: solutions/make_chocolate.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+def make_chocolate(small: int, big: int, goal: int) -> int:
+    """
```
Comment (PyaeLinn01) [inline solutions/make_chocolate.py:13]: The argument descriptions are clear, but mention explicitly that non-integer inputs are not supported (e.g., floats or strings).

File: solutions/make_chocolate.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    Determine the number of small bars needed to meet the goal weight.
+
+    Args:
+        small (int): The number of small bars available (1 kg each). Must be non-negative.
```
Comment (PyaeLinn01) [inline solutions/make_chocolate.py:17]: The return value is documented well. Consider adding more explanation about why -1 is returned when the goal cannot be met.

File: solutions/make_chocolate.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        big (int): The number of big bars available (5 kg each). Must be non-negative.
+        goal (int): The target weight of the chocolate package. Must be non-negative.
+
+    Returns:
+        int: The number of small bars needed to meet the goal.
```
Comment (PyaeLinn01) [inline solutions/make_chocolate.py:22]: The defensive check for non-negative inputs is good, but document in the docstring that this raises a ValueError.

File: solutions/make_chocolate.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+             Returns -1 if the goal cannot be achieved.
+
+    Raises:
```
Comment (PyaeLinn01) [inline solutions/make_chocolate.py:25]: The strategy is clear, with concise variable names like max_big_bars, big_bars_used, and remaining_goal. Add inline comments explaining why the minimum of max_big_bars and big is used.

File: solutions/tests/test_make_chocolate.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
+"""
```
Comment (PyaeLinn01) [inline solutions/tests/test_make_chocolate.py:1]: Ensure the test file name follows the convention /tests/test_make_chocolate.py to match the function file name.

File: solutions/tests/test_make_chocolate.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+Unit tests for the `make_chocolate` function in the `make_chocolate` module.
+
+These tests verify:
+- Correct calculation of required small bars
+- Defensive assertions for invalid inputs
+- Boundary cases where the goal cannot be achieved
+
+@author: May Mon Thant
```
Comment (PyaeLinn01) [inline solutions/tests/test_make_chocolate.py:9]: The module docstring is clear and describes the purpose of the tests. However, consider explicitly stating that edge cases, such as goal = 0, are included.

File: solutions/tests/test_make_chocolate.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+"""
+
+import sys
+import unittest
```
Comment (PyaeLinn01) [inline solutions/tests/test_make_chocolate.py:13]: The test class name TestMakeChocolate is descriptive and follows PascalCase. No changes needed.

File: solutions/tests/test_make_chocolate.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+from pathlib import Path
+
+# Add the parent directory to the system path to import the solutions module
```
Comment (PyaeLinn01) [inline solutions/tests/test_make_chocolate.py:16]: Each test name is clear and descriptive. However, add more detailed docstrings for tests that check specific logic or edge cases.

File: solutions/tests/test_make_chocolate.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        self.assertEqual(make_chocolate(0, 3, 15), 0)
+
+    def test_goal_unachievable_with_big_bars(self):
+        """Test when big bars alone cannot achieve the goal."""
+        self.assertEqual(make_chocolate(4, 2, 14), 4)
+
+    def test_goal_unachievable_no_small_bars(self):
+        """Test when no small bars are available to meet the goal."""
+        self.assertEqual(make_chocolate(0, 2, 11), -1)
+
```
Comment (PyaeLinn01) [inline solutions/tests/test_make_chocolate.py:48]: The code is well-structured and spaced to separate test cases clearly. No changes needed.

File: solutions/tests/test_make_chocolate.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    def test_exceeds_goal(self):
+        """Test when the goal exceeds the total available bars."""
+        self.assertEqual(make_chocolate(4, 1, 10), -1)
+
+    def test_mix_of_bars(self):
+        """Test a mix of big and small bars to meet the goal."""
+        self.assertEqual(make_chocolate(4, 1, 7), 2)
+
+    def test_goal_exact_with_big_bars(self):
+        """Test when the goal is met using only big bars."""
```
Comment (PyaeLinn01) [inline solutions/tests/test_make_chocolate.py:38]: Great job including boundary cases, such as goal = 0 and exact matches with bars. Consider adding cases for large inputs to test scalability.

File: solutions/tests/test_make_chocolate.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        self.assertEqual(make_chocolate(0, 3, 15), 0)
+
+    def test_goal_unachievable_with_big_bars(self):
+        """Test when big bars alone cannot achieve the goal."""
+        self.assertEqual(make_chocolate(4, 2, 14), 4)
+
```
Comment (PyaeLinn01) [inline solutions/tests/test_make_chocolate.py:44]: Excellent defensive assertion test. Include a similar test for non-integer input (e.g., strings or floats) if applicable.
